### PR TITLE
Correctly parse IPv6 addresses in Net::HTTP instrumentation

### LINKED ISSUE #2163: IPv6 Literal notation isn't handled correctly when making XMLRPC requests
### Issue Description

Hi - when I added Sentry to a rails app I started getting weird exceptions like this when using XMLRPC (NOTE: IPv6 works with basic Net::HTTP requests
```
irb(main):004:0>XMLRPC::Client.new('::1', '/RPC2', 8080).call('system.listMethods')
/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/rfc3986_parser.rb:67:in `split': bad URI(is not URI?): "http://::1/RPC2" (URI::InvalidURIError)
```

I've found with the backtrace it's caused by sentry hijacking any Net::HTTP requests for its own logging

```
/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/rfc3986_parser.rb:67:in `split': bad URI(is not URI?): "http://::1/RPC2" (URI::InvalidURIError)
"/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/rfc3986_parser.rb:67:in `split'",
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/rfc3986_parser.rb:72:in `parse'",              
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/common.rb:188:in `parse'",
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/net/http.rb:80:in `extract_request_info'",                                                  
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/net/http.rb:33:in `block in request'",                                                      
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/hub.rb:102:in `with_child_span'",                                                           
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry-ruby.rb:456:in `with_child_span'",                                                          
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/net/http.rb:32:in `request'",                                                               
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/net/http.rb:1484:in `request_post'",                                                                                                                            
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/xmlrpc-0.3.3/lib/xmlrpc/client.rb:498:in `do_rpc'",                                                                       
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/xmlrpc-0.3.3/lib/xmlrpc/client.rb:287:in `call2'",                                                                        
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/xmlrpc-0.3.3/lib/xmlrpc/client.rb:268:in `call'",   
```
I thought it was an issue with `xmlrpc` but it's not https://github.com/ruby/xmlrpc/issues/42

IPv6 literal notation uses square brackets to allow parsing of a port number:
`[::1]:8080`

```
sentry/net/http.rb:80 extract_request_info
        uri = req.uri || URI.parse("#{use_ssl? ? 'https' : 'http'}://#{address}#{req.path}")
```
This line tries to rebuild the URI from the attributes of the Net::HTTP instance, which fails because `address`  is `::1`. Not a valid hostname. I think you probably need to add a regex using the `resolv` gem or a manual check to see if the address contains colons meaning it's an IPv6 address.

The reason this works with basic Net::HTTP requests is when those are made they have their `uri` attribute set correctly. When making an `XMLRPC::Client` request, the uri isn't set so the fallback `URI.parse` is called which then fails

### Reproduction Steps

add sentry-ruby to your gemfile

```
require 'xmlrpc/client'
XMLRPC::Client.new('::1', '/RPC2', 8080).call('system.listMethods')
```

### Expected Behavior

XMLRPC call should work

### Actual Behavior

```
/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/rfc3986_parser.rb:67:in `split': bad URI(is not URI?): "http://::1/RPC2" (URI::InvalidURIError)
"/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/rfc3986_parser.rb:67:in `split'",
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/rfc3986_parser.rb:72:in `parse'",              
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/uri/common.rb:188:in `parse'",
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/net/http.rb:80:in `extract_request_info'",                                                  
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/net/http.rb:33:in `block in request'",                                                      
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/hub.rb:102:in `with_child_span'",                                                           
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry-ruby.rb:456:in `with_child_span'",                                                          
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/sentry-ruby-5.11.0/lib/sentry/net/http.rb:32:in `request'",                                                               
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/3.1.0/net/http.rb:1484:in `request_post'",                                                                                                                            
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/xmlrpc-0.3.3/lib/xmlrpc/client.rb:498:in `do_rpc'",                                                                       
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/xmlrpc-0.3.3/lib/xmlrpc/client.rb:287:in `call2'",                                                                        
 "/home/rowan/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/xmlrpc-0.3.3/lib/xmlrpc/client.rb:268:in `call'",   
```

### Ruby Version

3.1.0

### SDK Version

5.11

### Integration and Its Version

Rails + Sidekiq

### Sentry Config

Sentry.init do |config|
  config.enabled_environments = %w[production]
  config.dsn = Rails.configuration.sentry[:dsn]
  if Sidekiq.server?
    # We're in sidekiq
    # Don't use HTTP logger yet there's a bug that will kill CycleMatcher
    config.breadcrumbs_logger = [:sentry_logger, :http_logger]
  else
    # We're in the web app
    config.breadcrumbs_logger = [:active_support_logger, :http_logger]
  end

  # To activate performance monitoring, set one of these options.
  # We recommend adjusting the value in production:
  config.traces_sample_rate = 0.1
end

### PR DESCRIPTION
**Pull request by natikgadzhi**
## Summary

This pull request adds a `Net::HTTP` instrumentation fix for `extract_request_info` that makes it parse `IPv6` addresses correctly.

Fixes #2163. 

Thank you @Rotario for reporting and suggesting a fix!

## Changes
- Applied the workaround similar to what @Rotario suggested.
- Added a unit test.
- Added a changelog entry.

### CODE REVIEW

**Comment by natikgadzhi** [comment]
Work in progress. Seems like older Rubies do not like `Resolv`. I think I'll just grab the regex and put it in our net/http.rb.

File: sentry-ruby/spec/sentry/net/http_spec.rb [hunk 1]
```diff
@@ -9,6 +9,21 @@
...
+    end
+
+    it "correctly parses the short-hand IPv6 addresses" do
+      stub_normal_response
+
+      conn = Net::HTTP.new('::1', 8080)
+      expect do
+        conn.request_post('/path', 'foo=1')
+      end.not_to raise_error
+    end
```
Comment (st0012) [inline sentry-ruby/spec/sentry/net/http_spec.rb:24]: It's a bit nitpicking, but I still prefer to test things as much end-to-end as possible:

```suggestion
    it "correctly parses the short-hand IPv6 addresses" do
      stub_normal_response

      transaction = Sentry.start_transaction
      Sentry.get_current_scope.set_span(transaction)

      _ = Net::HTTP.get("::1", "/path", 8080)

      expect(transaction.span_recorder.spans.count).to eq(2)

      request_span = transaction.span_recorder.spans.last
      expect(request_span.data).to eq(
        { "url" => "http://[::1]/path", "http.request.method" => "GET", "http.response.status_code" => 200 }
      )
    end
```

**Review by st0012 — APPROVED** [review]
Thank you, it looks great 👍  I just have a nitpick on test.

File: sentry-ruby/lib/sentry/net/http.rb [hunk 2]
```diff
@@ -77,7 +77,10 @@ def from_sentry_sdk?
       end
 
       def extract_request_info(req)
-        uri = req.uri || URI.parse("#{use_ssl? ? 'https' : 'http'}://#{address}#{req.path}")
+        # IPv6 url could look like '::1/path', and that won't parse without
+        # wrapping it in square brackets.
+        hostname = address =~ Resolv::IPv6::Regex ? "[#{address}]" : address
```
Comment (st0012) [inline sentry-ruby/lib/sentry/net/http.rb:82]: I think you need `require "resolv"` to use it.

File: sentry-ruby/spec/sentry/net/http_spec.rb (hunk 1 above)
Comment (natikgadzhi) [reply sentry-ruby/spec/sentry/net/http_spec.rb:24]: Not at all — I think you're right, and I'm learning from you _a lot_ with each and every one of your comments and suggestions. Thank you for helping me learn!

Comment (st0012) [reply sentry-ruby/spec/sentry/net/http_spec.rb]: Thank you! I learn from your PRs too (such as this one) 😄

File: sentry-ruby/lib/sentry/net/http.rb (hunk 2 above)
Comment (natikgadzhi) [reply sentry-ruby/lib/sentry/net/http.rb:82]: Yup, checked in `irb`, you actually have to require things before using them. Who knew. 

Seems like on some Rubies, it gets required within `net/http`, but it's not guaranteed. Let's see if explicit require solves it!

File: sentry-ruby/lib/sentry/transport/http_transport.rb [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require "net/http"
+require "resolv"
```
Comment (st0012) [inline sentry-ruby/lib/sentry/transport/http_transport.rb:4]: Could you add it to the `net/http` patch instead as it's required by it, not `HTTPTransport`.

**Comment by natikgadzhi** [comment]
@st0012, cleaned things up! 
- Resolved the conflict
- Explicit require "resolv"
- Better scoped unit test (thank you!)

Let's see how CI gods feel about it.

File: sentry-ruby/lib/sentry/transport/http_transport.rb (hunk 3 above)
Comment (natikgadzhi) [reply sentry-ruby/lib/sentry/transport/http_transport.rb:4]: /facepalm my bad, one sec

Comment (natikgadzhi) [reply sentry-ruby/lib/sentry/transport/http_transport.rb]: Done.

File: sentry-ruby/spec/sentry/net/http_spec.rb [hunk 4]
```diff
@@ -9,6 +9,28 @@
     ::Logger.new(string_io)
   end
 
+  context "with IPv6 addresses" do
+    before do
+      perform_basic_setup
```
Comment (st0012) [inline sentry-ruby/spec/sentry/net/http_spec.rb:14]: Sorry that I forgot to mention with this approach you need to enable tracing here.

```suggestion
      perform_basic_setup do |config|
        config.traces_sample_rate = 1.0
      end
```

**Pull request merged**